### PR TITLE
Manchester | ITP-MAY-2026 | Yee Man Tsang | Sprint 2 | Coursework/Sprint-2

### PR DESCRIPTION
**Pull request by lintsang**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

Finished all exercises in all folders in sprint 2

### CODE REVIEW

File: Sprint-2/3-mandatory-implement/1-bmi.js [hunk 1]
```diff
@@ -15,5 +15,5 @@
 // It should return their Body Mass Index to 1 decimal place
 
 function calculateBMI(weight, height) {
-    // return the BMI of someone based off their weight and height
-}
\ No newline at end of file
+    return (weight / height**2).toFixed(1);
```
Comment (cjyuan) [inline Sprint-2/3-mandatory-implement/1-bmi.js]: What **type** of value do you expect your function to return? A number or a string?
Does your function return the **type** of value you expect?

Different types of values may appear identical in the console output, but they are represented and treated differently in the program. For example,
```javascript
  console.log(123);              // Output 123
  console.log("123");            // Output 123
  
  // Treated differently in the program
  let sum1 = 123 + 100;         // Evaluate to 223 -- a number
  let sum 2 = "123" + 100;      // Evaluate to "123100" -- a string.
```

File: Sprint-2/3-mandatory-implement/2-cases.js [hunk 2]
```diff
@@ -14,3 +14,8 @@
 // You will need to come up with an appropriate name for the function
 // Use the MDN string documentation to help you find a solution
 // This might help https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toUpperCase
+
+function UPPER_SNAKE_CASE(string){
+    upperString = string.toUpperCase();
```
Comment (cjyuan) [inline Sprint-2/3-mandatory-implement/2-cases.js]: Should `upperString` be declared using `const` or `let`?

File: Sprint-2/3-mandatory-implement/2-cases.js [hunk 3]
```diff
@@ -14,3 +14,8 @@
 // You will need to come up with an appropriate name for the function
 // Use the MDN string documentation to help you find a solution
 // This might help https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toUpperCase
+
+function UPPER_SNAKE_CASE(string){
```
Comment (cjyuan) [inline Sprint-2/3-mandatory-implement/2-cases.js]: Could you look up the naming conventions in JavaScript? In particulars,
- Variable and function names
- Class and Types names
- Named constants

Then, update the function name according to those conventions.

File: Sprint-2/5-stretch-extend/format-time.js [hunk 4]
```diff
@@ -1,11 +1,21 @@
 // This is the latest solution to the problem from the prep.
 // Make sure to do the prep before you do the coursework
 // Your task is to write tests for as many different groups of input data or edge cases as you can, and fix any bugs you find.
+function pad(num) {
+  let numString = num.toString();
+  while (numString.length < 2) {
+    numString = "0" + numString;
+  }
```
Comment (cjyuan) [inline Sprint-2/5-stretch-extend/format-time.js]: `num` is either a 1-digit or 2-digit number. A conditional statement is probably enough.

File: Sprint-2/5-stretch-extend/format-time.js [hunk 5]
```diff
@@ -1,11 +1,21 @@
...
+  return numString;
+}
 
 function formatAs12HourClock(time) {
   const hours = Number(time.slice(0, 2));
+  const minute = pad(Number(time.slice(3,5)));
```
Comment (cjyuan) [inline Sprint-2/5-stretch-extend/format-time.js]: - What's the difference between `minute` and `time.slice(3, 5)`?

File: Sprint-2/5-stretch-extend/format-time.js [hunk 6]
```diff
@@ -21,5 +31,33 @@ const currentOutput2 = formatAs12HourClock("23:00");
...
+
+const currentOutput4 = formatAs12HourClock("01:31");
+const targetOutput4 = "01:31 am";
+console.assert(
+  currentOutput4 === targetOutput4,
+  `current output: ${currentOutput4}, target output: ${targetOutput4}`
+);
+
+const currentOutput5 = formatAs12HourClock("00:00");
+const targetOutput5 = "00:00 am";
```
Comment (cjyuan) [inline Sprint-2/5-stretch-extend/format-time.js]: Shouldn't "00:00" be converted to "12:00 am"?

File: Sprint-2/5-stretch-extend/format-time.js [hunk 7]
```diff
@@ -1,11 +1,21 @@
...
   const hours = Number(time.slice(0, 2));
+  const minute = pad(Number(time.slice(3,5)));
   if (hours > 12) {
-    return `${hours - 12}:00 pm`;
+    return `${hours - 12}:${minute} pm`;
+  } else if (hours == 12 & minute == '00'){
+    return `${time} pm`;
   }
   return `${time} am`;
 }
```
Comment (cjyuan) [inline Sprint-2/5-stretch-extend/format-time.js:23]: What do you expect from the following function calls?
```
 formatAs12HourClock("08:35");
 formatAs12HourClock("20:35");
 formatAs12HourClock("12:34");
 formatAs12HourClock("00:34");
```

Do they return the values you expected? Are the returned values consistently formatted?

File: Sprint-2/3-mandatory-implement/1-bmi.js (hunk 1 above)
Comment (lintsang) [reply Sprint-2/3-mandatory-implement/1-bmi.js]: Thanks for reminder. I learnt that the types of value should be carefully considered when I am writing the code.

File: Sprint-2/3-mandatory-implement/2-cases.js (hunk 2 above)
Comment (lintsang) [reply Sprint-2/3-mandatory-implement/2-cases.js]: I would use const here as I am not going to reassign the value in this case.

File: Sprint-2/3-mandatory-implement/2-cases.js (hunk 3 above)
Comment (lintsang) [reply Sprint-2/3-mandatory-implement/2-cases.js]: Many thanks. I clearly understand the naming conventions in Javascript now.

File: Sprint-2/5-stretch-extend/format-time.js (hunk 4 above)
Comment (lintsang) [reply Sprint-2/5-stretch-extend/format-time.js]: Thanks for letting me know about this issue which I haven't noticed it before.

File: Sprint-2/5-stretch-extend/format-time.js (hunk 5 above)
Comment (lintsang) [reply Sprint-2/5-stretch-extend/format-time.js]: After reading the code again, I found the type of 'minute' is expected to be in string, while time.slice(3, 5) is already a string and I decided to assign to it directly. Thanks!

File: Sprint-2/5-stretch-extend/format-time.js (hunk 6 above)
Comment (lintsang) [reply Sprint-2/5-stretch-extend/format-time.js]: Yes, this is one of the conversion I haven't considered correctly.

File: Sprint-2/5-stretch-extend/format-time.js (hunk 7 above)
Comment (lintsang) [reply Sprint-2/5-stretch-extend/format-time.js:23]: Thanks for your suggestion. After reconsidering all this conversion, I think my code should be good to pass all 12 sample conversion check.

Comment (cjyuan) [reply Sprint-2/5-stretch-extend/format-time.js:23]: I only find 6 tests in the file and they are not comprehensive. Supposedly the function should work correctly for any valid argument, including those I mentioned in my previous comment.

**Comment by cjyuan** [comment]
Have you made any change based on the comments? If so, you would need to push them to GitHub so that I can review them.

File: Sprint-2/5-stretch-extend/format-time.js (hunk 7 above)
Comment (lintsang) [reply Sprint-2/5-stretch-extend/format-time.js:23]: Sorry I didn't know why there was only 6 tests before. But I created a new pull request and now I can see all 12 tests between the line 67 to 107. Please review again my change. I tested my code passed all tests. Thanks.

**Comment by cjyuan** [comment]
Changes look good.

I am not sure why GitHub shows me the older version of your files yesterday.

**Comment by illicitonion** [comment]
Closing PR because the May ITP run has finished. Feel free to re-open if you're still working on it.
